### PR TITLE
Client matchmaking

### PR DESCRIPTION
**Pull request by capturcus**
- choose between direct connection mode and matchmaking mode (and set default values) with an .ini file
- use a simple synchronous boost http server to connect to the matchmaker
- handle data received from the matchmaker
- minor fixes, including ui

### CODE REVIEW

File: client/dfclient/src/http_client.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	return res.result_int();
+}
+
+int http::MatchmakerGet(std::string address, std::string port, std::string& ret)
+{
+	int retCode = 0;
+	try {
+		retCode = throwingMatchmakerGet(address, port, ret);
+	} catch (...) {
+		return -1;
```
Comment (piodul) [inline client/dfclient/src/http_client.cpp:80]: How about printing the exception in the logs?

File: client/dfclient/src/http_client.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	// Set up an HTTP GET request message
+	bhttp::request<bhttp::string_body> req{bhttp::verb::get, "/matchmake", 10};
+	req.set(bhttp::field::host, address);
+	req.set(bhttp::field::user_agent, BOOST_BEAST_VERSION_STRING);
+
+	// Send the HTTP request to the remote host
+	bhttp::write(stream, req);
+
+	// This buffer is used for reading and must be persisted
```
Comment (piodul) [inline client/dfclient/src/http_client.cpp]: This comment is a little imprecise. Persisted, but for how long? What does "persisted" mean here?

File: client/dfclient/src/menu_state.cpp [hunk 3]
```diff
@@ -14,40 +18,119 @@
...
+	}
+}
+
+void MenuState::ProcessMatchmakerData(std::string data) {
+	boost::property_tree::ptree tree;
+	std::stringstream ss;
+	ss << data;
+	boost::property_tree::read_json(ss, tree);
+
+	auto status = tree.get("status", "");
```
Comment (piodul) [inline client/dfclient/src/menu_state.cpp:104]: Is the server-matchmaker protocol documented somewhere?

File: client/dfclient/src/websocket.cpp [hunk 4]
```diff
@@ -163,16 +163,20 @@ int WebSocketBeast::Connect(std::string& fullAddress) {
...
+			[](websocket::request_type& req)
+			{
+				req.set(http::field::user_agent,
+					std::string(BOOST_BEAST_VERSION_STRING) +
+						" websocket-client-coro");
+			}));
+		ws.handshake(host, "/");
+	} catch (std::exception e) {
+		return -1;
+	}
```
Comment (piodul) [inline client/dfclient/src/websocket.cpp:179]: What's the reason for this change?

File: client/dfclient/src/http_client.cpp (hunk 2 above)
Comment (capturcus) [reply client/dfclient/src/http_client.cpp]: Actually read the comments copied from an example. Removed the "persisted" part. Everything else checks out

File: client/dfclient/deadfish.ini.example [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[default]
+client_mode=matchmaker
+matchmaker_address=localhost
+matchmaker_port=8000
+direct_address=localhost
+direct_port=63987
```
Comment (mparafin) [inline client/dfclient/deadfish.ini.example:6]: The viable options are not documented anywhere. Maybe some comments here could be a good idea, as it is an example file? But anywhere would be ok, even in the [README.md](https://github.com/capturcus/deadfish/blob/master/README.md).

File: client/dfclient/include/menu_state.hpp [hunk 6]
```diff
@@ -16,7 +20,20 @@ struct MenuState
...
+	void DirectConnectLayout();
+
+	ClientMode clientMode = ClientMode::Matchmaker;
+	std::string matchmakerAddress = "localhost";
+	std::string matchmakerPort = "8000";
+	std::string directAddress = "localhost";
+	std::string directPort = "63987";
+
+	char buf0[64];
+	char buf1[64];
```
Comment (mparafin) [inline client/dfclient/include/menu_state.hpp]: Is there no way to use less non-descriptive names? :D They seem to be indentical in functionality and one of them  is always a nickname, while second is __sometimes__ a server adress. Maybe they could be assigned a single role and named as such.

File: client/dfclient/src/menu_state.cpp [hunk 7]
```diff
@@ -14,40 +18,119 @@
...
+{
+	nc::SceneNode &rootNode = nc::theApplication().rootNode();
+	auto res = nc::theApplication().appConfiguration().resolution;
+	auto text = new ncine::TextNode(&rootNode, _resources.fonts["comic"].get());
+	text->setString(message.c_str());
+	text->setPosition(res.x * 0.5f, res.y * 0.30f);
+	text->setScale(0.5f);
+	text->setColor(0, 0, 0, 255);
+	_resources._tweens.push_back(CreateTextTween(text));
+}
```
Comment (mparafin) [inline client/dfclient/src/menu_state.cpp]: I still don't like the idea of leaking TextNodes just to show them xd I know it's copied, just nitpicking. Maybe I'll refactor it someday. :D

**Review by mparafin — APPROVED** [review]
L(V)GTM :) You have my approve, all the comments are just nits of the kind of code hygiene. For me the code is mergeable even as it is :)

File: client/dfclient/src/menu_state.cpp (hunk 3 above)
Comment (capturcus) [reply client/dfclient/src/menu_state.cpp:104]: It is now in the README

File: client/dfclient/src/websocket.cpp (hunk 4 above)
Comment (capturcus) [reply client/dfclient/src/websocket.cpp:179]: Adding a try/catch block serves to return a value on error instead of crashing with an uncaught exception

File: client/dfclient/deadfish.ini.example (hunk 5 above)
Comment (capturcus) [reply client/dfclient/deadfish.ini.example:6]: Added coments in the ini example file

File: client/dfclient/include/menu_state.hpp (hunk 6 above)
Comment (capturcus) [reply client/dfclient/include/menu_state.hpp]: Renamed buffers and assigned roles to them

File: client/dfclient/src/menu_state.cpp (hunk 7 above)
Comment (capturcus) [reply client/dfclient/src/menu_state.cpp]: True, I don't know why did I commit this, sorry. Added deleting the text pointer when the tween reaches 0
